src/wrapper_array.jl

```diff
@@ -0,0 +1,81 @@
+
+struct NamedDimsArray{L<:Tuple, T, N, A<:AbstractArray{T,N}} <: AbstractArray{T,N}
+    data::A
+end
+
+function NamedDimsArray(orig::AbstractArray{T,N}, names) where {T, N}
+    if length(names) != N
+        throw(ArgumentError("A $N dimentional array, need $N dimension names. Got: $names"))
+    end
+    names_tt = Tuple{names...}
+    return NamedDimsArray{names_tt, T, N, typeof(orig)}(orig)
+end
+
+
+Base.parent(x::NamedDimsArray) = x.data
+
+
+"""
+    dim_names(A)
+
+Returns a tuple of containing the names of all the dimensions of the array `A`.
+"""
+dim_names(::Type{<:NamedDimsArray{L}}) where L = Tuple(L.parameters)
+dim_names(x::T) where T = dim_names(T)
+
+
+name2dim(a::NamedDimsArray{L}, name) where L = name2dim(L, name)
+
+
+
+#############################
+# AbstractArray Interface
+# https://docs.julialang.org/en/v1/manual/interfaces/index.html#man-interface-array-1
+
+Base.size(a::NamedDimsArray) = size(parent(a))
+Base.getindex(A::NamedDimsArray, inds...) = getindex(parent(A), inds...)
+Base.setindex(A::NamedDimsArray, value, inds...) = setindex(parent(A), value, inds...)
+
+
+###############################
+# kwargs indexing
+
+"""
+    order_named_inds(A, named_inds...)
+
+Returns the indices that have the names and values given by `named_inds`
+sorted into the order expected for the dimension of the array `A`.
+If any dimensions of `A` are not present in the named_inds,
+then they are given the value `:`, for slicing
+
+For example:
+```
+A = NamedDimArray(rand(4,4), (:x,, :y))
+order_named_inds(A; y=10, x=13) == (13,10)
+order_named_inds(A; x=2, y=1:3) == (2, 1:3)
+order_named_inds(A; y=5) == (:, 5)
+```
+
+This provides the core indexed lookup for `getindex` and `setindex` on the Array `A`
+"""
+function order_named_inds(A; named_inds...)
+    keys(named_inds) ⊆ dim_names(A) || throw(
+        DimensionMismatch("Expected $(dim_names(A)), got $(keys(named_inds))")
+    )
+
+
+    inds = map(dim_names(A)) do name
+        get(named_inds, name, :)  # default to slicing
+    end
+end
```

Comment:
Alternative from @iamed2 

Something that does
```
defaults = (a=:, b=:, c=:)  # In the right order
Tuple(merge(defaults, named_inds))
```

So we would need a function that generates that default.

@iamed2  says this would give compile time sorting,
though I am not sure it would with the Error we throw.
Compile time **hates** things that can throw exceptions.

Reply:
If the error means bail out, compile time is fine with it. I believe it's only a problem when you're catching exceptions or when the exception can change the output.

Reply:
I will have to recheck tgat

Reply:
Seems right

```
julia> @code_llvm (()->NamedDims.order_named_inds((:a,)))()

;  @ none:1 within `#41'
define nonnull %jl_value_t addrspace(10)* @"japi1_#41_14567"(%jl_value_t addrspace(10)*, %jl_value_t addrspace(10)**, i32) #0 {
top:
  %3 = alloca %jl_value_t addrspace(10)**, align 8
  store volatile %jl_value_t addrspace(10)** %1, %jl_value_t addrspace(10)*** %3, align 8
  ret %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 5019490584 to %jl_value_t*) to %jl_value_t addrspace(10)*)
}

julia> @code_llvm (()->NamedDims.order_named_inds((:a,), b=2))()

;  @ none:1 within `#43'
; Function Attrs: noreturn
define nonnull %jl_value_t addrspace(10)* @"japi1_#43_14568"(%jl_value_t addrspace(10)*, %jl_value_t addrspace(10)**, i32) #0 {
top:
  %3 = alloca %jl_value_t addrspace(10)**, align 8
  store volatile %jl_value_t addrspace(10)** %1, %jl_value_t addrspace(10)*** %3, align 8
  %4 = call nonnull %jl_value_t addrspace(10)* @"julia_#order_named_inds_14569"({ i64 } addrspace(11)* nocapture readonly addrspacecast ({ i64 }* @0 to { i64 } addrspace(11)*), %jl_value_t addrspace(10)* addrspacecast (%jl_value_t* inttoptr (i64 4595989792 to %jl_value_t*) to %jl_value_t addrspace(10)*))
  call void @llvm.trap()
  unreachable
}
```